pkg/gofr/datasource/file/ftp/fs.go

```diff
@@ -0,0 +1,358 @@
+func (f *fileSystem) MkdirAll(name string, _ os.FileMode) error {
+	defer f.processLog(&FileLog{Operation: "MkdirAll", Location: path.Join(f.config.RemoteDir, name)}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("MkdirAll failed. Provide a valid path : %v", errEmptyPath)
+		return errEmptyPath
+	}
+	// returns a slice of those directories that do not exist with the first index being the latest existing parent directory path.
+	dirs := f.mkdirAllHelper(name)
+
+	currentDir := dirs[0]
+
+	for i, dir := range dirs {
+		if i == 0 {
+			continue
+		}
+
+		currentDir = path.Join(currentDir, dir)
+
+		err := f.conn.MakeDir(currentDir)
+		if err != nil {
+			f.logger.Errorf("MkdirAll failed : %v", err)
+
+			return err
+		}
+	}
+
+	f.logger.Logf("Directories creation completed successfully.")
+
+	return nil
+}
+
+// Open retrieves a file from the FTP server and returns a file handle.
+func (f *fileSystem) Open(name string) (container.File, error) {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Open", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Open_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return nil, errEmptyFilename
+	}
+
+	res, err := f.conn.Retr(filePath)
+	if err != nil {
+		f.logger.Errorf("Open_file failed. Error opening file : %v", err)
+		return nil, err
+	}
+
+	defer res.Close()
+
+	filename := path.Base(filePath)
+
+	f.logger.Logf("Open_file successful. Filepath : %q", filePath)
+
+	return &file{
+		response: res,
+		name:     filename,
+		path:     filePath,
+		conn:     f.conn,
+		logger:   f.logger,
+		metrics:  f.metrics,
+	}, nil
+}
+
+// permissions are not clear for Ftp as file commands do not accept an argument and don't store their file permissions.
+// currently, this function just calls the Open function.
+// Here, os.FileMode is unused, but is added to comply with FileSystem interface.
+func (f *fileSystem) OpenFile(name string, _ int, _ os.FileMode) (container.File, error) {
+	return f.Open(name)
+}
+
+// Remove deletes a file from the FTP server.
+func (f *fileSystem) Remove(name string) error {
+	filePath := path.Join(f.config.RemoteDir, name)
+
+	defer f.processLog(&FileLog{Operation: "Remove", Location: filePath}, time.Now())
+
+	if name == "" {
+		f.logger.Errorf("Remove_file failed. Provide a valid filename : %v", errEmptyFilename)
+		return errEmptyFilename
+	}
+
+	err := f.conn.Delete(filePath)
+
+	if err != nil {
+		var textprotoError *textproto.Error
+
+		if errors.As(err, &textprotoError) && textprotoError.Msg == transferCompleteError.Msg {
+			f.logger.Logf("Remove_file successful. File with path %q successfully removed", filePath)
+			return nil
+		}
+
+		f.logger.Errorf("Remove_file failed. Error while deleting the file: %v", err)
+
+		return err
+	}
+
+	f.logger.Logf("Remove_file success. File with path %q successfully removed", filePath)
+
+	return nil
+}
```

Comment:
I would suggest this to avoid having two branches doing the exact sane thing

```suggestion
  var textprotoError *textproto.Error
  switch {
    // expected error
   case errors.As(err, &textprotoError) && textprotoError.Code == textproto.StatusClosingDataConnection && whatever check on the message:
    // nothing to do

    case err != nil:
		  f.logger.Errorf("Remove_file failed. Error while deleting the file: %v", err)
		  return err
	}

	f.logger.Logf("Remove_file success. File with path %q successfully removed", filePath)

	return nil
}
```